### DISCUSSION TITLE
[20 Marzo 2023] Modifica CPU

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#69] [Modifiche architettura MIPS (M-Z)]
![Esame](https://github.com/Elia-Belli/ae2223/assets/107944476/c29f914d-699e-446f-ab42-aca6023dcc97)
![CPU](https://github.com/Elia-Belli/ae2223/assets/107944476/35e07461-b557-4749-a247-810648631214)
---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11666079/AE.2021-22.-.Appello.S2.-.2023-03-20.-.Esercizio.CPU.-.Scheda.di.calcolo.pdf)

**Comment by Elia-Belli** [comment]
## 1) Architettura Modificata
![CPU](https://github.com/sapienzastudentsnetwork/archelab2223/assets/107944476/dd36c3e0-ac04-4657-a09f-a40c96391e09)

**Comment by daromacarlo** [comment]
![prima](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/d0956cef-aacf-4a9a-afad-71ff3584e693)
![secondù](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/cbe7cd5b-e856-4104-80cd-5babdc1b6ab3)
![third](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/b1ccf79e-a96d-4050-87f2-b9404c4ade67)

piccola svista nell'esercizio 3. il periodo di clock deve essere allungato dato che la nuova istruzione è più lenta rispetto a LoadWord che era l'istruzione più lenta prima della modifica dell'architettura, in particolare il periodo di clock deve essere 750ns, quindi lafrequenza 1/750ns.

**Comment by FeddyLix17** [comment]
## 1. Modifiche all'architettura

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/82747e84-4962-4b70-8783-24e6d345ad66)

il disegno è praticamente analogo a quello di @Elia-Belli se non per il fatto che invece di implementare un ulteriore shifter utilizzo quello presente sul lato superiore del disegno

## 2. Word che esprime l'istruzione *lwig $s4, $t4, 0x0203*

ricordando che
- L'Opcode dell'istruzione lwig (fornito in esadecimale dalla traccia) è uguale a 0x3D, in binario 00111101, codificato in 6 bit 111101
- il campo rs, rappresentate il primo registro "sorgente", è $s4, codificato in 5 bit come 10100 (registro numero 20)
- il campo rt, rappresentante il secondo registro "sorgente", è $t0, codificato in 5 bit come 01100 (registro numero 12)
- il campo immediate, rappresentante il valore costante con cui bisognerà lavorare, è uguale a 0x0203, codificato in 16 bit come 0000001000000011

si ottiene la sequenza completa

| posizione bit della word  | 31 | 30 | 29 | 28 | 27 | 26 | 25 | 24 | 23 | 22 | 21 | 20 | 19 | 18 | 17 | 16 | 15 | 14 | 13 | 12 | 11 | 10 | 9 | 8 | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 |
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| campo istruzione | \| | | Op | code | | \| | \| | rs | = | s4 | \| | \| | rt | = | t0 | \| |  \| |  |  | | | | immediate | | | | | | | | | \| |
| sequenza effettiva | 1 | 1 | 1 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 0 | 0 | 1 | 1 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 1 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 1 | 1 |

## 3. Durata totale ciclo di Clock
per calcolare la durata totale del ciclo di clock si analizzano i passaggi effettuati dalla nuova istruzione
1. si accede alla memoria istruzioni in fase di istruction fetch, con un tempo (fornito dall'esercizio) di 200 ns
2. si accede ai registri in fase di istruction decode, con un tempo (fornito dall'esercizio) di 50 ns
3. l'ALU calcola il val2_address in fase di Execution , con un tempo (fornito dall'esercizio) di 150 ns
4. si accedede alla memoria dati per estrarre il valore contenuto all'indirizzo val2_address, con un tempo sempre di 200 ns
5. viene calcolato il valore più grande tra $val1 e $val2 usando un sommatore che come l'ALU ha tempo 150 ns
6. si accede nuovamente ai registri per trascrivere il valore più grande tra i 2 nel registro $outreg, con un tempo sempre di 50 ns

| IF | 200 ns | \| | ID | 50 ns | \| | EXE | 150 ns | \| | MEM | 200 ns | \| | EXE | 150 ns | \| | WB | 50 ns |
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| PC + 4  | 150 ns | \| | | | | | | | | | | | | | |

si conclude che per poter ammettere l'esecuzione della nuova istruzione la durata totale del ciclo di clock dovrà essere uguale a  200 + 50 + 150 + 200 + 150 + 50 = 800 ns

## 4. valore assegnato a $v0 al termine dell'esecuzione del codice
non ho veramente idea di come si risolva, @CarloDaRomadev tu come hai fatto?

**Comment by daromacarlo** [comment]
@FedVlogger17  ciao,  se noti errori nei miei ragionamenti o hai domande scrivimi pure :)
![Cattu33333333333333333333333333ra](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/ba44967b-3a45-4008-b641-795e512316eb)
![Cattueeeeeeeeeeeeeeeeeeeeeeeeeeeeera](https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/b520651f-bba6-4313-aad4-36c859f6f1b6)

**Comment by FeddyLix17** [comment]
@CarloDaRomadev ti ringrazio per la spiegazione del tuo ragionamento, dopo averla letta ed aver riletto con attenzione la domanda numero 4 dell'esercizio questo è ciò che ho concluso:

- Se la cpu è difettosa tale che RegDst sia uguale a 1 - Jump, vuol dire che RegDst sarà sempre uguale a 1 tranne per l'istruzione j Exit (che poco importa visto che il datapath di una istruzione J non prevede la l'accesso ai registri)
- ci si aspetta che le 2 istruzioni "rotte" saranno quelle di tipo I, perchè il valore immediato sarà codificato nei bit [15-0] ed il mux avente come selettore RegDst andrà comunque a selezionare i bit [15:11] come bit del registro di destinazione, salvando potenzialmente i dati non si sa dove

definito ciò, ho analizzato (come richiesto dall'esercizio) il particolare comportamento di quel blocco di codice
- nell'istruzione addi $s0, $s0, 0x8004; 0x8004 rappresenta la sequenza 1000000000000100, ovvero i bit [15:0], quindi la sequenza [15:11] sarà uguale a 10000, ovvero il registro numero 16, fortuna vuole che sia proprio $s0
- l'istruzione sub $s1, $s0, $zero opera normalmente e quindi $s1=$s0 - 0 = 0x8004
- l'istruzione beq $s1, $s0, Out effettuerà il branch all'etichetta Out in quanto in quel momento $s1 e $s0 sono uguali
- verrà quindi eseguita correttamente l'istruzione add $v0, $0, $s1, quindi $v0 = $s1 = 0x8004
- nell'istruzione addi $v0, $v0, 0x8002; infine, 0x8002 rappresenterà la sequenza 1000000000000010, ed i bit [15:11] saranno 10000, quindi sempre nel registro $s0 andremo ad aggiungere il valore immediato invece che in $v0

secondo quest'ipotesi, alla fine dell'esecuzione del codice avremo
- $s0 = 0x8004 + 0x8002 = 0x10006 (65542 in decimale)
- $s1 = 0x8004 (32772 in decimale)
- $v0 =  0x8004

**Comment by CasuFrost** [comment]
![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/97389762/db32e88e-ba7d-47e7-ad19-ea05fb6528a0)
![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/97389762/cedd0173-b792-4311-82dc-e20f8f90d995)
![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/97389762/2612f392-efc5-4bde-a817-4e50514f85be)

**Comment by daromacarlo** [comment]
La seconda istruzione salva in $zero la differenza tra $s0 e $zero, mi sono accorto di questo errore

**Comment by Frisbon** [comment]
@CasuFrost Grazie per la spiegazione dettagliata dell'esercizio 4, la cosa dei bit che scrive comunque su $s0 mi ha completamente inculato

**Comment by daromacarlo** [comment]
@FeddyLix17 @Frisbon @CasuFrost @Elia-Belli Ho interpretato male il punto 4 scusate.
